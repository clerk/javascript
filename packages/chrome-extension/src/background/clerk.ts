import { Clerk } from '@clerk/clerk-js';

import {
  createClerkClient as _createClerkClient,
  type CreateClerkClientOptions as _CreateClerkClientOptions,
} from '../internal';
import { SCOPE } from '../types';

export type CreateClerkClientOptions = Omit<_CreateClerkClientOptions, 'scope'>;

export let clerk: Clerk;

export async function createClerkClient(opts: CreateClerkClientOptions): Promise<Clerk> {
  if (clerk) {
    return clerk;
  }

  Clerk.mountComponentRenderer = undefined;
  clerk = await _createClerkClient({ ...opts, scope: SCOPE.BACKGROUND });
  await clerk.load({ standardBrowser: false });

  return clerk;
}
