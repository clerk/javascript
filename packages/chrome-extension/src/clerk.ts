import { Clerk } from '@clerk/clerk-js';
import { DEV_BROWSER_JWT_KEY } from '@clerk/shared';
import { parsePublishableKey } from '@clerk/shared/keys';
import browser from 'webextension-polyfill';

import { AUTH_HEADER, CLIENT_JWT_KEY, DEFAULT_LOCAL_HOST_PERMISSION } from './constants';
import type { ClerkClientExtensionFeatures } from './types';
import type { GetClientCookieParams } from './utils/cookies';
import { assertPublishableKey } from './utils/errors';
import { JWTHandler } from './utils/jwt-handler';
import { getValidPossibleManifestHosts, validateHostPermissionExistence, validateManifest } from './utils/manifest';
import { BrowserStorageCache, type StorageCache } from './utils/storage';

export let clerk: Clerk;

Clerk.sdkMetadata = {
  name: PACKAGE_NAME,
  version: PACKAGE_VERSION,
};

type ClerkClientOptions = {
  extensionFeatures?: ClerkClientExtensionFeatures;
  publishableKey: string;
  storageCache?: StorageCache;
};

export async function createClerkClient({
  extensionFeatures = {},
  publishableKey,
  storageCache = BrowserStorageCache,
}: ClerkClientOptions): Promise<Clerk> {
  if (clerk) {
    return clerk;
  }

  // Parse publishableKey and assert it's present/valid, throw if not
  const key = parsePublishableKey(publishableKey);
  assertPublishableKey(key);

  const isProd = key.instanceType === 'production';
  const manifest = browser.runtime.getManifest();

  // Will throw if manifest is invalid
  validateManifest(manifest, extensionFeatures);

  let jwt: JWTHandler | undefined;

  if (extensionFeatures.sync) {
    const hostHint = isProd ? key.frontendApi : DEFAULT_LOCAL_HOST_PERMISSION;
    const validHosts = getValidPossibleManifestHosts(manifest);

    // Will throw if manifest host_permissions doesn't contain a valid host
    validateHostPermissionExistence(validHosts, hostHint);

    // Set up cookie params based on environment
    const getClientCookieParams: GetClientCookieParams = isProd
      ? {
          urls: `https://${key.frontendApi}`,
          name: CLIENT_JWT_KEY,
        }
      : {
          urls: validHosts,
          name: DEV_BROWSER_JWT_KEY,
        };

    // Set up JWT handler and attempt to get JWT from storage on initialization
    jwt = JWTHandler(storageCache, { ...getClientCookieParams, frontendApi: key.frontendApi, sync: true });
  } else {
    jwt = JWTHandler(storageCache, { frontendApi: key.frontendApi, sync: false });
  }

  if (!jwt) {
    throw new Error('StorageCache could not be initialized.'); // TODO: Update error
  }

  // Create Clerk instance
  clerk = new Clerk(publishableKey);

  if (extensionFeatures.background) {
    Clerk.mountComponentRenderer = undefined;
  }

  // Append appropriate query params to all Clerk requests
  clerk.__unstable__onBeforeRequest(async requestInit => {
    requestInit.credentials = 'omit';

    const currentJWT = await jwt.get();

    console.log('__unstable__onBeforeRequest currentJWT', currentJWT);

    if (!currentJWT) {
      requestInit.url?.searchParams.append('_is_native', '1');
      return;
    }

    if (isProd) {
      requestInit.url?.searchParams.append('_is_native', '1');
      (requestInit.headers as Headers).set('authorization', `Bearer ${currentJWT}`);
    } else {
      requestInit.url?.searchParams.append('__clerk_db_jwt', currentJWT);
    }
  });

  // Store updated JWT in StorageCache on Clerk responses
  clerk.__unstable__onAfterResponse(async (_, response) => {
    const authHeaderkey = isProd ? AUTH_HEADER.production : AUTH_HEADER.development;
    const authHeader = response?.headers.get(authHeaderkey);

    if (authHeader?.startsWith('Bearer')) {
      const newJWT = authHeader.split(' ')[1] || undefined;

      if (newJWT) {
        await jwt.set(newJWT);
      } else {
        await jwt.remove();
      }
    } else if (authHeader) {
      await jwt.set(authHeader);
    }
  });

  if (extensionFeatures.background) {
    await clerk.load({ standardBrowser: false });
  }

  return clerk;
}
