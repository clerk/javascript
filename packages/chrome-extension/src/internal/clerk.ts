import { Clerk } from '@clerk/clerk-js/no-rhc';
import { DEV_BROWSER_JWT_KEY } from '@clerk/shared/devBrowser';
import { parsePublishableKey } from '@clerk/shared/keys';
import browser from 'webextension-polyfill';

import { SCOPE, type Scope } from '../types';
import { CLIENT_JWT_KEY, DEFAULT_LOCAL_HOST_PERMISSION } from './constants';
import { assertPublishableKey } from './utils/errors';
import { JWTHandler } from './utils/jwt-handler';
import { validateManifest } from './utils/manifest';
import { requestHandler } from './utils/request-handler';
import { responseHandler } from './utils/response-handler';
import { BrowserStorageCache, type StorageCache } from './utils/storage';

export let clerk: Clerk;

Clerk.sdkMetadata = {
  name: PACKAGE_NAME,
  version: PACKAGE_VERSION,
};

export type CreateClerkClientOptions = {
  publishableKey: string;
  scope?: Scope;
  storageCache?: StorageCache;
  syncHost?: string;
};

export async function createClerkClient({
  publishableKey,
  scope,
  storageCache = BrowserStorageCache,
  syncHost,
}: CreateClerkClientOptions): Promise<Clerk> {
  if (scope === SCOPE.BACKGROUND) {
    Clerk.mountComponentRenderer = undefined;
  }

  // Don't cache background scripts as it can result in out-of-sync client information.
  if (clerk && scope !== SCOPE.BACKGROUND) {
    return clerk;
  }

  // Sync is enabled if a `syncHost` is provided
  const sync = Boolean(syncHost);

  // Parse publishableKey and assert it's present/valid, throw if not
  const key = parsePublishableKey(publishableKey);
  assertPublishableKey(key);

  const isProd = key.instanceType === 'production';
  const manifest = browser.runtime.getManifest();

  // Will throw if manifest is invalid
  validateManifest(manifest, {
    background: scope === SCOPE.BACKGROUND,
    sync,
  });

  // Set up JWT handler and attempt to get JWT from storage on initialization
  const url = syncHost ? syncHost : DEFAULT_LOCAL_HOST_PERMISSION;

  const jwtOptions = {
    frontendApi: key.frontendApi,
    name: isProd ? CLIENT_JWT_KEY : DEV_BROWSER_JWT_KEY,
    sync,
    url,
  };

  const jwt = JWTHandler(storageCache, jwtOptions);

  // Create Clerk instance
  clerk = new Clerk(publishableKey);
  clerk.__unstable__onAfterResponse(responseHandler(jwt, { isProd }));
  clerk.__unstable__onBeforeRequest(requestHandler(jwt, { isProd }));

  return clerk;
}
