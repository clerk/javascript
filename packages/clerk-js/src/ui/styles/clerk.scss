@import '../../../../shared/components/styles/variables';
@import '../../../../shared/components/styles/mixins';

// postcss is responsible for adding a cl- prefix to all declarations
// Look at the webpack config for how that happens

$cl-box-shadow-lg: 0px 10px 20px rgba(60, 60, 60, 0.1);
$cl-black: #000000;
$cl-gray: #4e4e4e;
$cl-gray-light-1: #eeeeee;
$cl-gray-light-2: #828282;
$cl-gray-light-3: $themed-font-color-l1;
$cl-gray-light-4: #757575;
$cl-red: rgba(205, 68, 74, 1);
$cl-red-O10: rgba(205, 68, 74, 0.1);

// Scope everything to cl-mount-node
.component {
  @import '../../../../shared/components/styles/variables';
  @import '../../../../shared/components/styles/breakpoints';
  @import '../../../../shared/components/styles/themed_mixins';
  @import '../../../../shared/components/styles/mixins';

  @import 'injections';
  @import 'custom_reboot';
  @import 'sharedComponentsOverwrites';
  @import 'shared/fonts';
  @import 'shared/logo';
  @import 'shared/poweredByClerk';
  @import 'theme';
  @import 'authForms';
  @import 'backButton';
  @import 'error';
  @import 'userprofile';
  @import 'userbutton';
  @import 'backButton';
  @import 'activeAccountsManager';
  @import 'signInAccountSwitcher';
  @import 'verifyMagicLink';

  font-family: $theme-font-family;

  *,
  *:before,
  *:after {
    box-sizing: inherit;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }

  :focus:not(:focus-visible) {
    outline: none !important;
  }

  :focus-visible {
    @include outline();
  }

  fieldset {
    border: none;
  }
}

.modal-backdrop {
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: rgba(0, 0, 0, 0.2);
  z-index: $high-z-index;
}

.modal-container {
  padding: 0.5rem;
  display: flex;
  max-width: 100%;
  max-height: 100%;
  height: 100%;
  width: 100%;
  z-index: $high-z-index;
  border-radius: 0;
  background-color: rgba(0, 0, 0, 0.2);

  .sign-in,
  .sign-up {
    margin: 0;
  }
}
