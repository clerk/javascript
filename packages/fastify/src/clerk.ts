import { Clerk } from '@clerk/clerk-sdk-node';

import { API_KEY, API_URL, API_VERSION, JWT_KEY, SECRET_KEY } from './constants';

export const clerkClient = Clerk({
  apiKey: API_KEY,
  secretKey: SECRET_KEY,
  apiUrl: API_URL,
  apiVersion: API_VERSION,
  jwtKey: JWT_KEY,
});
