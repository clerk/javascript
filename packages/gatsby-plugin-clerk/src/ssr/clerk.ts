import { Clerk } from '@clerk/backend';

export const API_URL = process.env.CLERK_API_URL || 'https://api.clerk.dev';
export const API_VERSION = process.env.CLERK_API_VERSION || 'v1';
export const API_KEY = process.env.CLERK_API_KEY || '';
export const SECRET_KEY = process.env.CLERK_SECRET_KEY || '';
export const FRONTEND_API = process.env.GATSBY_CLERK_FRONTEND_API || '';
export const PUBLISHABLE_KEY = process.env.GATSBY_CLERK_PUBLISHABLE_KEY || '';

const clerkClient = Clerk({
  apiKey: API_KEY,
  secretKey: SECRET_KEY,
  apiUrl: API_URL,
  apiVersion: API_VERSION,
  // TODO: Fetch version from package.json
  userAgent: 'gatsby-plugin-clerk',
});

const createClerkClient = Clerk;

export { clerkClient, createClerkClient, Clerk };

export * from '@clerk/backend';
