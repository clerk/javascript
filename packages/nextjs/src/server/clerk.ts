import { Clerk } from '@clerk/backend';

export const API_URL = process.env.CLERK_API_URL || 'https://api.clerk.dev';
export const API_VERSION = process.env.CLERK_API_VERSION || 'v1';
export const API_KEY = process.env.CLERK_API_KEY || '';
export const SECRET_KEY = process.env.CLERK_SECRET_KEY || '';
export const FRONTEND_API = process.env.NEXT_PUBLIC_CLERK_FRONTEND_API || '';
export const PUBLISHABLE_KEY = process.env.NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY || '';

const clerkClient = Clerk({
  apiKey: API_KEY,
  secretKey: SECRET_KEY,
  apiUrl: API_URL,
  apiVersion: API_VERSION,
  // TODO: Fetch version from package.json
  userAgent: '@clerk/nextjs',
});

const createClerkClient = Clerk;

export { clerkClient, createClerkClient, Clerk };

export * from '@clerk/backend';
