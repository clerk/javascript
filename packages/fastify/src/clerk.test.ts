const ClerkMock = jest.fn(() => {
  return 'clerkClient';
});
jest.mock('@clerk/clerk-sdk-node', () => {
  return { Clerk: ClerkMock };
});

import { clerkClient } from './clerk';

describe('clerk', () => {
  afterAll(() => {
    jest.resetModules();
  });

  test('initializes clerk with constants', () => {
    expect(ClerkMock.mock.calls).toMatchSnapshot();
    expect(clerkClient).toEqual('clerkClient');
  });
});
