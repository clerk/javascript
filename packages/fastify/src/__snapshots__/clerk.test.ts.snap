// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clerk initializes clerk with constants 1`] = `
[
  [
    {
      "apiUrl": "https://api.clerk.dev",
      "apiVersion": "v1",
      "jwtKey": "",
      "secretKey": "TEST_API_KEY",
    },
  ],
]
`;
